README.md
# The Brass Ring

This repository is used to hold files based on the Dead Souls mudlib (http://dead-souls.net) used to develop The Brass Ring mud.

NOTE: These files may change radically!

Included is the DikuMud to Dead Souls port (https://github.com/LashMUD/DikuMud-to-Dead-Souls-Port) and the approprite licenses for each MUD.
This will be the main repository for the Diku to Dead port which will no longer be updated.


changelog.txt
Changelog (year-month-day)
22-12-29
*changed directory structure for wyr
 added more rooms and npc's

*added all skills to be added for level change
 - need 15 skill points for level change per level
   1st level need 15 total skill points
   10th level need 150 total skill points and so on

*/lib/domains/town/npc/dirk.c
 - added leveling based on skills to dirk.c

*/lib/domains/etnar/wyr/wyr/npc/albert/albert_derby
 - added code for npc's getting rumors from a list in a file

*/lib/domains/diku-alfa/room/30.zon/npc/3024_sorcerer_guard
 - added npc spell use to sorcerer in mages guild in Midgaard

*/lib/include/combat_messages.h
 - changed combat messages so hit type is bolded red for player attacking

*/lib/secure/cmds/players
 - moved mudtime command so players can use it

*/lib/cmds/players
 - skills command changed so Minor Skills is shown - not Other Skills

*/lib/lib/body.c
 - modified int GetHealRate() function for sleeping buff

*/lib/lib/player.c
 - changed SetClass() to adventurer for skill based leveling
   there is no "adventurer" class just used for aesthetics
 - changed "Drifter" to "Adventurer" in commands and other files to be consistent:
   /lib/secure/daemon/finger.c
   /lib/secure/obj/medtric.c
   /lib/cmds/players/kills.c
   /lib/cmds/players/skills.c
   /lib/cmds/players/stat.c 
   /lib/cmds/players/stats.c 

*/lib/lib/props/rent.c
 - created rent property so player can rent rooms at Inns and get healing buffs 

*/lib/secure/connect.c
 - changed so that if classselection and skillselection are in use
   the player can choose to use one of the pregenerated classes in
   /lib/secure/cfg/classes or select their own skills

*/lib/secure/cfg/classes
 - added bard class for Fostaine Pyre npc in wyr
 - redid all classes so they can be used for skill system advancement

*/lib/doc/help/skills
 - added help files for skills

2020-06-25
*/lib/lib/genetics.c 
 - changed variables in eventBlind() function to emulate previous code from
   original DS mudlib

2020-06-18
*/lib/lib/player.c
 - added code so player is not still blinded or cursed after death regeneration
*/lib/lib/props/factions.c
 - modified AddFactions function so can add multiple factions through the 
   function

2020-06-07
*/lib/lib/npc.c
 - changed SetActionsMap() functionality so it works as intended
 - added varargs type to eventRestoreSight()
 - fixed a typo in the changelog (!)

2020-06-02
*/lib/lib/bonus.c
 - added mappings for functionality
 - file flows better and the code is more robust
 - apologies to Cratylus for my previous attempts and totally
   butchering this file
*/lib/lib/potion.c
 - added mappings for functionality
*/lib/lib/genetics.c
 - changed code in function eventRestoreSight(object who, int amt) 
   so sight can be restored immediately. Put in a check if 'amt' is not given
   as an argument
*/lib/domains/Etnar/Wyr/meals/4050_purple_potion.c
 - changed code to fit with above file modifications

2020-05-28
*/lib//lib/props/factions.c
 - recoded so that it works as intended - redid SetFactions() function
*/domains/town/npc/leo.c
 - used new SetFactions() function
*/lib/doc/The_Brass_Ring
 - rewrote some of this help file - still needs work

2020-05-21
* several files have had minor editing changes related to file headers,
  addition of file headers if not present, and in file comments removal that
  were more of a distraction than a help 
*/lib/secure/cfg preload.cfg
 - remove from repo - mud specific
*/lib/secure/daemon/autexec.c 
 - remove from repo - mud specific
*/lib/lib/props/factions.c 
 -removed SetFactions() function - was not initialized properly and not required
*/lib/lib/genetics.c 
 - removed file revisions (restructuring) and distracting comments
*/lib/power/npc/include/magic_user.h
 -removed eventCurse() function declaration - not required since the actual curse
  function is inherited
*/lib/powers/npc/curse.c
 -added checks to make sure that if an object was cursed they could not be cursed again
  if a user used the curse function without proper formatting the object cursed
  would have the cursed flag removed but the skill or stat modification would
  remain   

2020-05-16
-added comments to some files for information
 and coding clarifications/explanations

2020-05-12
*/lib/lib/genetics.c
 -rearranged file contents so function structures flow together
 -added comments about changes
*/lib/secure/sefun/light.c
 -added line to mixed_check_light function to let player know if
  they can't see due to being blinded
*/lib//lib/lvs/level.c added comments
*/lib/lib/lvs/abilities.c
  -added comments and GetOtherSkills() function
  -ResetLevel() used in player.c during player loading
   if commented out throws an error when player logging in

2020-05-07
*/lib/secure/include/global.h and /lib/secure/cfg/global_template.cfg 
 -changed DAY_ONE to 1449853200 Dec. 11, 2015 12:00:00
   date The Brass Ring was uploaded to GitHub (Fri Dec 11 12:00:00 2015)

lib/lib/player.c
/*    /lib/player.c
 *    from the Dead Souls LPC Library
 *    the player object
 *    created by Descartes of Borg 950321
 *    Version: @(#) player.c 1.26@(#)
 *    Last Modified: 97/01/03
 *    
 *    modified by lash (ccoker) for use in
 *    The Brass Ring mud
 *     - added code for factions used in TBR
 *     - player should not be cursed or blinded after death regeneration
 *     - added timer checking for room renting
 *     last modified: 22/11/18 year/month/day
 */

#include <lib.h>
#include ROOMS_H
#include <damage_types.h>
#include <message_class.h>
#include <daemons.h>
#include <vision.h>
#include <position.h>
#include "include/player.h"

#undef LIB_PLAYER_SKILL_LOSS
#define PERCENT_XP 0.02
#define PERCENT_SP 0.70
#define PERCENT_MP 0.95
#define PERCENT_HP 0.70

inherit LIB_INTERACTIVE;
inherit LIB_LIVING;
inherit LIB_FACTIONS;
inherit LIB_RENT;

private string *Titles;
string *Muffed = ({});
private mapping *Deaths;
private int TrainingPoints, TitleLength;
private static int heartcount;
private mixed telopt_ip;

/* *****************  /lib/player.c driver applies  ***************** */

static void create(){
    interactive::create();
    living::create();

    Titles = ({});
    TitleLength = 1;
}

static void heart_beat(){
    mixed heartping;
    int idle;
    int rented;

    if(!interactive(this_object())) return;

    heartping = GetProperty("keepalive");
    idle = query_idle(this_object());
    heartcount++;

    if(GetProperty("reply")){
        int t = GetProperty("reply_time");
        if(t > 1 && (time() - t) > 90000){
            RemoveProperty("reply");
            RemoveProperty("reply_time");
        }
    }
    factions::CheckTimer();        
    rent::CheckTimer();
    interactive::heart_beat();
    living::heart_beat();

    if( IDLE_TIMEOUT && idle >= IDLE_TIMEOUT 
            && !builderp(this_object()) 
            && !present("testchar badge",this_object()) 
            && !present("idler_amulet",this_object()) 
            && !testp(this_object()) ){
        cmdQuit();
        return;
    }
#ifdef __DSLIB__
    if(heartping){
        send_nullbyte(this_object());
    }
#endif
}

static void net_dead(){
    interactive::net_dead();
    set_heart_beat(0);
}

static int Destruct(){
    CHARACTER_D->eventSaveTime();
    living::Destruct();
    return interactive::Destruct();
}

mixed eventAsk(object who, string what){
    if( what != "convert me" ) return 0;
    if( !GetReligion(1) ) return 0;
    who->SetProperty("converting", GetReligion(1));
    return 1;
}

void eventKillEnemy(object ob){
    living::eventKillEnemy(ob);
    STATISTICS_D->eventKill(ob);
}

void eventReconnect(){
    interactive::eventReconnect();
    living::eventReconnect();
    set_heart_beat(GetHeartRate());
}

/************** player.c command functions *************/
varargs int eventShow(object who, string str){
    if( !living::eventShow(who, str) ) return 0;
    if( this_player() != this_object() )
        eventPrint(this_player()->GetName() + " looks you over.");
    return 1;
}

/* *****************  /lib/player.c events  *************** */

varargs mixed eventDisplayStatus(int simple){
    string str;
    int qp, xp, hp, mp, sp, max_hp, max_mp, max_sp;

    hp = GetHealthPoints();
    max_hp = GetMaxHealthPoints();
    mp = GetMagicPoints();
    max_mp = GetMaxMagicPoints();
    sp = GetStaminaPoints();
    max_sp = GetMaxStaminaPoints();
    xp = GetExperiencePoints();
    qp = GetQuestPoints();

    if( percent(hp, max_hp) < 20.0 )
        str = "%^YELLOW%^hp: %^RED%^" + hp + "%^RESET%^/" + max_hp;
    else str = "%^YELLOW%^hp: %^RESET%^" + hp + "/" + max_hp;
    if( percent(mp, max_mp) < 20.0 )
        str += "  %^BLUE%^mp: %^RED%^" + mp + "%^RESET%^/" + max_mp;
    else str += "  %^BLUE%^mp: %^RESET%^" + mp + "/" + max_mp;
    if( percent(sp, max_sp) < 20.0 )
        str += "  %^GREEN%^sp: %^RED%^" + sp + "%^RESET%^/" + max_sp;
    else str += "  %^GREEN%^sp: %^RESET%^" + sp + "/" + max_sp;
    if(!simple){
        str += "  %^MAGENTA%^xp: %^RESET%^" + xp;
        str += "  %^CYAN%^qp: %^RESET%^" + qp;
    }
    if(!simple){
        message("status", str, this_object());
        return 1;
    }
    return str;
}

static void eventDestroyUndead(object agent){
}

varargs int eventDie(mixed agent){
    int x;
    string agentname;

    this_object()->SetDying(0);

    if(!agent) agent = previous_object();
    if(!agent) agent = this_object();
    if(stringp(agent)) agentname = agent;
    else agentname = agent->GetName();

    if( (x = living::eventDie(agent)) != 1 ){
        return x;
    }
    if(!Deaths || !sizeof(Deaths)) 
        Deaths = ({([ "date" : ctime(time()), "enemy" : agentname ])});
    else Deaths += ({ ([ "date" : ctime(time()), "enemy" : agentname ]) });

    if( !GetUndead() ){
        eventDestroyUndead(agent);
    }
    else  {
        message("my_action", "Consciousness passes from you after one last "
                "gasp for air.", this_object());
        message("my_action", "You awake, but you find your body feels "
                "different, and the world about you is unfamiliar.",
                this_object());
        if( agent ){
            message("other_action", GetName() + " is killed by "
                    + agentname + ".",
                    environment(this_object()), ({ agent, this_object() }));
            message("other_action", "You send " + GetName() + " into the "
                    "Underworld.", agent);
        }
        else message("other_action", GetName() + " dies.",
                environment(), ({ this_object() }) );

        NewBody(GetRace());

        eventCompleteHeal(GetMaxHealthPoints()/2);
        AddMagicPoints(-(random(GetMagicPoints())));
        this_object()->eventMove(ROOM_DEATH);
        this_object()->save_player(this_object()->GetKeyName());
        this_object()->eventDescribeEnvironment();
    }
    flush_messages();
    return 1;
}

mixed eventTurn(object who){
    if( !living::eventTurn(who) ){
        return 0;
    }
    eventDestroyUndead(who);
    return 1;
}

varargs void eventRevive(int nopenalty){
    string skill;

    this_object()->SetDead(0);
    if(!creatorp(this_object())) this_object()->SetInvis(0);
    this_object()->SetDeathEvents(0);

    if( !GetUndead() ) return;
    SetUndead(0);
    if(this_player()->GetPoison() > 0){
        this_player()->AddPoison(0 - this_player()->GetPoison());
    }
    if(!nopenalty && newbiep(this_object())) {
        nopenalty = 1;
        write("As a newbie you don't incur an experience penalty"
                " for this death.\n");
    }
    if(!nopenalty){
        int expee, subexpee;
#ifdef LIB_PLAYER_SKILL_LOSS
        foreach(skill in GetSkills()){
            int x;

            if( !random(4) ){
                continue;
            }
            if( newbiep(this_object()) ){
                x = 2;
            }
            else {
                x = 10;
            }
            x = random(x - (2*GetSkillClass(skill)))/2;
            if( x > 0 ){
                while( x-- ){
                    AddSkillPoints(skill,
                            -GetMaxSkillPoints(skill,
                                GetBaseSkillLevel(skill)));
                }
            }
        }
#endif
        expee = this_object()->GetExperiencePoints();
        subexpee = to_int(expee * PERCENT_XP);
        this_object()->AddExperienceDebt(subexpee);
    }
    NewBody(GetRace());
    eventCompleteHeal(GetMaxHealthPoints());
    SetSpecialTarget( ({}) );
    AddMagicPoints(-(GetMaxMagicPoints() * PERCENT_MP));
    AddStaminaPoints(-(GetMaxStaminaPoints() * PERCENT_SP));
    AddHealthPoints(-(GetMaxHealthPoints() * PERCENT_HP));
    if(this_object()->GetLead()){
        int shots = this_object()->GetLead("gunshot_wounds");
        if(shots) this_object()->AddLead("gunshot_wounds", -shots);
        shots = this_object()->GetLead("rifleshot_wounds");
        if(shots) this_object()->AddLead("rifleshot_wounds", -shots);
    }
    if(this_object()->GetCursed()) this_object()->RemoveCurse();
    if(this_object()->GetBlind()) this_object()->eventRestoreSight();
    if(creatorp()){
        string livingtitle = this_object()->GetLivingShort();
        if(!livingtitle) livingtitle = "$N the reborn";
        interactive::SetShort(livingtitle);
    }
}

int eventMove(mixed dest){
    int ret;
    object env = environment();
    string location;

    if(!env) location = ROOM_START;
    else if(clonep(env)) location = file_name(env);
    else location = base_name(env);

    if(location) this_object()->SetProperty("LastLocation", location);

    ret = interactive::eventMove(dest);
    if( this_object() && environment(this_object())) eventMoveFollowers(environment(this_object()));
    return ret;
}

int eventReleaseObject(object foo){
    object ob;

    ob = previous_object();
    if( !ob || !interactive::eventReleaseObject() ) return 0;
    if( ob->GetMass() ){
        AddCarriedMass( -(ob->GetMass()) );
        if(environment()) environment()->AddCarriedMass(-(ob->GetMass()));
    }
    return 1;
}

void eventLoadObject(mixed *value, int recurse){ }

/* *****************  /lib/player.c modal functions  ***************** */

int CanReceive(object ob){ return CanCarry(ob->GetMass()); }

mixed CanUse(){ return 1; }

/* *****************  /lib/player.c local functions  ***************** */

int Setup(){
    string classes;
    string oldparties = PARTY_D->GetOldParties();
    if( !interactive::Setup() ) return 0;
    /* changed by lash (ccoker) for skill_advance option */
    if( !GetClass() && CLASS_SELECTION ) SetClass("adventurer");
    RemoveExtraChannels(oldparties);
    foreach(string oldparty in oldparties){
        RemoveChannel(oldparty);
    }
    RemoveExtraChannels(oldparties);
    if( GetClass() ){
        foreach(classes in CLASSES_D->GetClasses())
            if( ClassMember(classes) && classes != GetClass() )
                AddChannel(classes);
    }
    if(sizeof(GetExtraChannels())) AddChannel(GetExtraChannels());
    set_heart_beat(GetHeartRate());
    if(builderp()){
        AddChannel( ({ "builder" }) );
        AddSearchPath( ({ DIR_BUILDER_CMDS, DIR_SECURE_BUILDER_CMDS }) );
    }
    PLAYERS_D->CheckBuilder(this_object());

    if(GetProperty("brand_spanking_new")){
        object jeans, shirt, book;
        if(!(this_object()->GetProperty("screen reader"))){
            this_object()->SetProperty("minimapping", 1);
        }
        else this_object()->SetProperty("minimapping", 0);
        if(ENGLISH_ONLY) this_object()->SetNativeLanguage("English");
        PLAYERS_D->AddPlayerInfo(this_object());

        foreach(classes in CLASSES_D->GetClasses())
            if( ClassMember(classes) && classes != GetClass() )
                AddChannel(classes);
        if( avatarp() ) AddChannel(({ "avatar" }));
        AddChannel( ({ "gossip", "newbie" }) );
        if( high_mortalp() ) AddChannel( ({ "hm" }) );
        else {
            //RemoveChannel( ({ "newbie" }) );
        }
        AddChannel( ({ "gossip" }) );
        if( councilp() ) AddChannel( ({ "council" }) );
        /* changed by lash or error thrown */
        if( GetClass() ) AddChannel(GetClass());

        jeans = new("/domains/default/armor/jeans");
        shirt = new("/domains/default/armor/shirt");
        book = new("/domains/default/obj/handbook");

        if(jeans) jeans->eventMove(this_object());
        if(shirt) shirt->eventMove(this_object());
        if(book && !present("handbook",this_object()))  
            book->eventMove(this_object());
        else if(book) book->eventMove(ROOM_FURNACE);

        if(jeans) this_object()->eventForce("wear jeans");
        if(shirt) this_object()->eventForce("wear shirt");
        SetProperty("brand_spanking_new",0);
    }

    if(this_object()->GetTown() == "FirstAdmin"){
        object robe, hat, staff, book, book2;
        string home;

        this_object()->SetTown("World");
        this_object()->SetProperty("minimapping", 0);
        if(!(this_object()->GetProperty("screen reader"))){
            this_object()->SetProperty("wizmapping", 1);
        }
        else this_object()->SetProperty("wizmapping", 0);
        robe = new("/domains/default/armor/robe");
        hat = new("/domains/default/armor/wizard_hat");
        staff = new("/secure/obj/staff");
        book = new("/domains/default/obj/guide");
        book2 = new("/domains/default/obj/handbook");

        if(robe) robe->eventMove(this_object());
        if(hat) hat->eventMove(this_object());
        if(staff) staff->eventMove(this_object());
        if(book) book->eventMove(this_object());
        if(book2) book2->eventMove(this_object());

        if(robe) this_object()->eventForce("wear robe");
        if(hat) this_object()->eventForce("wear hat");

        home = "/realms/"+this_player()->GetKeyName()+"/workroom";

        if(file_exists(home+".c")) 
            this_object()->eventMoveLiving(home);

        this_object()->AddChannel( ({"admin", "error", "cre", "newbie", "gossip", "ds", "ds_test", "lpuni", "death", "connections","intercre","dchat"}) );

        SetShort("First Admin $N");
    }
    if(!creatorp(this_object())) this_object()->SetInvis(0);
    RemoveProperty("reply");
    RemoveProperty("reply_time");
    return 1;
}

/* ***************** /lib/player.c data functions  ***************** */

int AddCurrency(string type, int amount){
    if( currency_value(amount, type) > 999 )
        log_file("currency", GetCapName() + " received "+amount+" "+type+
                " "+ctime(time())+"\n"+identify(previous_object(-1))+"\n");
    return living::AddCurrency(type, amount);
}

int AddBank(string bank, string type, int amount){
    if( currency_value(amount, type) > 999 )
        log_file("bank", GetCapName() + " deposited "+amount+" "+type+
                " "+ctime(time())+" into bank: "+bank+"\n" +
                identify(previous_object(-1))+"\n");
    return living::AddBank(bank, type, amount);
}

string *GetMuffed(){
    return Muffed;
}

string *SetMuffed(string *muffed){
    if(sizeof(muffed)) Muffed = muffed;
    else Muffed = ({});
    return Muffed;
}

string *AddMuffed(string muffed){
    string tmpstr;
    if(!muffed || muffed == "" || !sizeof(muffed)) return Muffed;
    if(muffed[0..0] == "@"){
        tmpstr = INTERMUD_D->GetMudName(muffed[1..]);
    }
    if(sizeof(tmpstr)) muffed = tmpstr;
    muffed = lower_case(muffed);
    if(member_array(muffed,Muffed) == -1) Muffed += ({ muffed });
    return Muffed;
}

string *RemoveMuffed(string unmuffed){
    string tmpstr;
    if(!sizeof(unmuffed)) return Muffed;
    if(grepp(unmuffed,"@")){
        tmpstr = INTERMUD_D->GetMudName(unmuffed[1..sizeof(unmuffed)-1]);
    }
    if(sizeof(tmpstr)) unmuffed = tmpstr;
    unmuffed = lower_case(unmuffed);
    if(member_array(unmuffed,Muffed) != -1) Muffed -= ({ unmuffed });
    return Muffed;
}

string *SetTitles(string *titles){
    if( sizeof(distinct_array(titles)) != sizeof(titles) ) return Titles;
    Titles = titles;
    SetShort("whatever");
}

string *AddTitle(mixed title){
    if(arrayp(title)){
        if(this_object()->GetGender() == "female") title = title[1];
        else title = title[0];
    }
    if( !stringp(title) ) return Titles;
    else if( member_array(title, Titles) != -1 ) return Titles;
    else {
        Titles = ({ title }) + Titles;
        SetShort("whatever");
        return Titles;
    }
}

string *RemoveTitle(mixed title){
    if( stringp(title) ) title = ({ title });
    foreach(string sub in title){
        Titles -= ({ sub });
        SetShort("whatever");
    }
    return Titles;
}

string *GetTitles(){ return Titles; }
int SetTitleLength(int x){
    if( x > 2 ) x = 2;
    return (TitleLength = x);
}

int GetTitleLength(){ return TitleLength; }

string SetShort(string irrelevant){
    string title, tmp;
    int i;

    if( high_mortalp() ) title = "High mortal ";
    else title = "";
    if(title && title != "") title += " $N ";
    else title = "$N ";
    if( GetUndead() && (tmp = GetUndeadType()) )
        return interactive::SetShort(title + "the " + tmp);
    if( avatarp() || creatorp() ){
        return interactive::SetShort(irrelevant);
    }
    if( !(i = sizeof(Titles)) )
        return interactive::SetShort(title + "the unaccomplished");
    else title += Titles[0];
    if( i > 1 && TitleLength > 1 ) title += " and " + Titles[1];
    return interactive::SetShort(title);
}

string GetPlainShort(){
    return interactive::GetShort();
}

int SetUndead(int x){
    x = living::SetUndead(x);
    SetShort("nonsense");
    return x;
}

string GetName(){
    if(GetInvis() && (!this_player() ||!this_player()->GetWizVision())){
        return "A shadow";
    }
    else return interactive::GetName();
}

varargs string GetLong(string str){
    mapping counts;
    string item;
    string *affects = ({});

    str = this_object()->GetShort() + "\n";
    str += interactive::GetLong() + "\n";
    str += living::GetLong(nominative(this_object()));
    foreach(item in map(all_inventory(),
                (: $1->GetAffectLong(this_object()) :))){
        if(item && member_array(item,affects) == -1) affects += ({ item });
    }
    if(sizeof(affects)) str += implode(affects,"\n")+"\n";
    if(this_object()->GetAffectLong()) str += this_object()->GetAffectLong();
    counts = ([]);
    foreach(item in map(
                filter(all_inventory(), (: !($1->GetInvis(this_object())) :)),
                (: $1->GetEquippedShort() :)))
        if( item ) counts[item]++;
    if( sizeof(counts) ) str += GetCapName() + " is carrying:\n";
    foreach(item in keys(counts))
        str += capitalize(consolidate(counts[item], item)) + "\n";
    return str;
}

string GetCapName(){ return interactive::GetCapName(); }

int ResetLevel(){
    int x, y;

    x = GetLevel();
    if( x != (y = living::ResetLevel()) ){
        string file;

        if( x > y ) file = "decline";
        else file = "advance";
        log_file(file, GetCapName() + " went from level " + x + " to "
                "level " + y + " (" + ctime(time()) + ")\n");
        if( x < y ){
            eventPrint("%^YELLOW%^You are now a more experienced " + 
                    GetClass() + ".");
            TrainingPoints += ( (y-x) * 4 );
        }
        else TrainingPoints -= ( (x-y) * 4 );
        SetShort("whatever");
        if( x > 49 && y < 50 ) RemoveChannel("avatar");
        if( x > 24 && y < 25 ) RemoveChannel("hm");
    }
    return y;
}

string SetClass(string str){
    if( GetClass() != living::SetClass(str) ){
        int points = TrainingPoints;
        string classes;

        ResetLevel();
        TrainingPoints = points;   /* leave points alone */
        AddChannel(GetClass());
        foreach(classes in CLASSES_D->GetClasses())
            if( ClassMember(classes) && classes != GetClass() )
                AddChannel(classes);
    }
    return GetClass();
}

varargs mixed GetEffectiveVision(mixed location, int raw_score){
    if( newbiep(this_object()) ) return VISION_CLEAR;
    else if(raw_score && location) return living::GetEffectiveVision(location,raw_score);
    else if(location) return living::GetEffectiveVision(location);
    else return living::GetEffectiveVision();
}

varargs static int AddHealthPoints(int x, string limb, object agent){
    int hp, ret, undead;
    hp = GetHealthPoints();
    undead = GetUndead();
    ret = living::AddHealthPoints(x, limb, agent);
    if( hp <= (hp = GetHealthPoints()) || undead != GetUndead() ) return ret;
    eventDisplayStatus();
    return ret;
}

int GetLanguageLevel(string lang){
    if( newbiep() ) return 100;
    else return living::GetLanguageLevel(lang);
}

mapping *GetDeaths(){
    if( !Deaths ) return ({});
    return copy(Deaths);
}

int AddTrainingPoints(int x){
    log_file("TrainingPoints", GetName() + " received " + x + " training "
            "points at " + ctime(time()) + "\ncall chain: " +
            sprintf("%O\n", previous_object(-1)) );
    return (TrainingPoints += x);
}

int RemoveTrainingPoints(int x){
    return (TrainingPoints -= x);
}

int GetTrainingPoints(){ return TrainingPoints; }

varargs int eventTrain(string skill, int points){
    float x = 0;
    mapping mp;

    if( points < 1 ) points = 1;
    if( !(mp = GetSkill(skill)) ) return 0;
    if( TrainingPoints < points ) return 0;
    while( points-- ){
        int max = GetMaxSkillPoints(skill, mp["level"]);
        switch( mp["class"] ){
            case 1: x = 50.0; break;
            case 2: x = 40.0; break;
            case 3: x = 30.0; break;
            case 4: x = 20.0; break;
            default: return 0;
        }
        TrainingPoints--;
        AddSkillPoints(skill, to_int( (max * x) / 100 ));
    }
    return 1;
}

mixed GetTeloptIp(){
    if(previous_object() && base_name(previous_object()) != SEFUN &&
            !(archp(this_player())) && previous_object() != this_object()) return 0;
    return telopt_ip;
}

mixed SetTeloptIp(mixed str){
    if(base_name(previous_object()) != LIB_CONNECT &&
            !(archp(this_player()))) return 0;
    return telopt_ip = str;
}
